Add array precision options for mwa_corr_fits files

plaplant opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
This change is motivated by #826, and gives the user the opportunity to select data types for the `data_array` and `nsample_array` when reading in MWA corr_fits files. The previous hard-coded defaults of `np.complex128` and `np.float32`, respectively, consumed more memory than typically necessary, especially for instances where no manipulation of these arrays was being done (e.g., when RFI flagging with SSINS).

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. If this PR closes an issue, put the word 'closes' before the issue link to auto-close the issue when the PR is merged. -->
Part of the motivation for this change is that the data for MWA corr_fits files is saved at single precision (`np.complex64`), so the previous default was artificially increasing the precision. Though subsequent calculations (e.g., applying calibration solutions) would benefit from this increased precision, analysis steps that only required reading the values without modifying them would not.

For the nsample array, the option of half-precision floats (`np.float16`) is added, though its use is discouraged (and mentioned in the docstrings). Floating-point round-off errors can be relatively large (~1e-3), but for raw corr_fits data that contains only 1s and 0s, there is no loss of precision. Care should be taken when averaging or down-selecting (e.g., LST-binning) to ensure that the calculation is being done at sufficient precision.


## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation change (documentation changes only)
- [ ] Version change
- [ ] Build or continuous integration change


## Checklist:
<!--- You may remove the checklists that don't apply to your change type(s) or just leave them empty -->
<!--- Go over all the following points, and replace the space with an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [contribution guide](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/.github/CONTRIBUTING.md).
- [x] My code follows the code style of this project.

New feature checklist:
- [x] I have added or updated the docstrings associated with my feature using the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] I have updated the tutorial to highlight my new feature (if appropriate).
- [x] I have added tests to cover my new feature.
- [x] All new and existing tests pass.
- [x] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/CHANGELOG.md).

plaplant commented:
@cjordan I've made this branch (`mwa_corr_fits_precision`) to implement the changes in your original PR with some added flexibility. You should now be able to do something like this:
```
uv.read_mwa_corr_fits(filelist, data_array_dtype=np.complex64, nsample_array_dtype=np.float16)
```
Let me know if that gives you the memory footprint you had before. Also note that these options will have to be propagated to modules that are reading these files, like SSINS.

bhazelton commented on pyuvdata/uvdata/uvdata.py:
```diff
@@ -5904,8 +5904,8 @@ def read_mwa_corr_fits(
         phase_to_pointing_center=False,
-        phase_data=None,
-        phase_center=None,
+        data_array_dtype=np.complex64,
```
Why is this defaulted differently than the `read` method or the `read_mwa_corr_fits` method on the MWACorrFITS object?

plaplant replied:
Good question. I think because I wasn't paying attention :)

I think given the raw data is single precision, I'd advocate for the default here and in the `MWACorrFITS.read_mwa_corr_fits` method to match that (`np.complex64`), but keep the default on the `UVData.read()` double-precision to avoid changing the current default on users. What are your thoughts?

bhazelton replied:
That sounds reasonable to me.

bhazelton commented:
@plaplant should we consider adding some options to the averaging methods to promote these arrays?

plaplant commented:
@bhazelton that's an excellent suggestion re: datatype promotion. I'll go though and try to find all the points where we should do that.

plaplant commented:
@bhazelton I added handling of the `nsample_array` precision in the `downsample_in_time` function, but I think that's the only place in `uvdata.py` that actually does arithmetic with it (rather than just down-selecting). I also added a test for it and some in-line comments.

bhazelton commented:
I think it's also averaged in `frequency_average`. I should also note that the logic in `downsample_in_time` is changed in PR #821, so whichever is merged second will need to do a careful rebase.

bhazelton commented on CHANGELOG.md:
```diff
@@ -5,9 +5,11 @@ All notable changes to this project will be documented in this file.
 
 ### Added
 - `propagate_flags` keyword for `UVData.frequency_average` which flags averaged samples if any contributing samples were flagged
+- `nsample_array_dtype` keyword for `UVData.read`, which sets the datatype for the nsample_array. Currently only used for mwa_corr_fits files.
```
This should be at the top -- reverse chronological order.

plaplant replied:
Good catch, thanks! Should be fixed now.

bhazelton commented on CHANGELOG.md:
```diff
@@ -5,9 +5,11 @@ All notable changes to this project will be documented in this file.
 
 ### Changed
 - `utils.uvcalibrate` now incorporates many more consistency checks between the uvcal and uvdata object. The new keywords `time_check` and `ant_check` were added to control some of these checks.
+- `data_array_dtype` keyword for `UVData.read` is now also respected by mwa_corr_fits files. Previously it was only used by uvh5 files.
```
same comment here

cjordan commented:
Hey - I don't want to derail any useful conversation here, just want to say thanks for the work here. I haven't had time to test it through SSINS yet, but I'll do that when I can (hopefully over the next couple of days).

plaplant commented:
@bhazelton I think this PR is ready to be looked at again when you get a chance.

bhazelton commented on pyuvdata/uvdata/uvdata.py:
```diff
@@ -6004,9 +6024,9 @@ def read_mwa_corr_fits(
         end_flag=2.0,
         flag_dc_offset=True,
         phase_to_pointing_center=False,
-        phase_data=None,
-        phase_center=None,
```
Good catch on these!
